README.md
# 🌅 Twilight

**Twilight** is a Lua plugin for Neovim 0.5 that dims inactive portions of the code you're editing.

The plugin was heavily inspired by [Limelight](https://github.com/junegunn/limelight.vim),
but uses [TreeSitter](https://github.com/nvim-treesitter/nvim-treesitter) for better dimming.

Pairs well with [zen-mode](https://github.com/folke/zen-mode.nvim).

![image](https://user-images.githubusercontent.com/292349/125419804-051321c2-d040-41c8-93fc-834b5f1098e3.png)

## ⚡️ Requirements

- Neovim >= 0.5.0

## 📦 Installation

Install the plugin with your preferred package manager:

### [packer](https://github.com/wbthomason/packer.nvim)

```lua
-- Lua
use {
  "folke/twilight.nvim",
  config = function()
    require("twilight").setup {
      -- your configuration comes here
      -- or leave it empty to use the default settings
      -- refer to the configuration section below
    }
  end
}
```

### [vim-plug](https://github.com/junegunn/vim-plug)

```vim
" Vim Script
Plug 'folke/twilight.nvim'

lua << EOF
  require("twilight").setup {
    -- your configuration comes here
    -- or leave it empty to use the default settings
    -- refer to the configuration section below
  }
EOF
```

## ⚙️ Configuration

Twilight comes with the following defaults:

```lua
{
  dimming = {
    alpha = 0.25, -- amount of dimming
    -- we try to get the foreground from the highlight groups or fallback color
    color = { "Normal", "#ffffff" },
    term_bg = "#000000", -- if guibg=NONE, this will be used to calculate text color
    inactive = false, -- when true, other windows will be fully dimmed (unless they contain the same buffer)
  },
  context = 10, -- amount of lines we will try to show around the current line
  treesitter = true, -- use treesitter when available for the filetype
  -- treesitter is used to automatically expand the visible text,
  -- but you can further control the types of nodes that should always be fully expanded
  expand = { -- for treesitter, we we always try to expand to the top-most ancestor with these types
    "function",
    "method",
    "table",
    "if_statement",
  },
  exclude = {}, -- exclude these filetypes
}
```

## 🚀 Usage

You can manually trigger **Twilight** using the commands:

- `Twilight`: toggle twilight
- `TwilightEnable`: enable twilight
- `TwilightDisable`: disable twilight

## 🧘 Zen Mode Integration

If you have [zen-mode](https://github.com/folke/zen-mode.nvim) installed, then **Twilight**
is activated automatically. Refer to the Zen Mode documentation to disable.


CHANGELOG.md
# Changelog

## 1.0.0 (2023-01-04)


### Features

* added config option for default bg when transparent [#20](https://github.com/folke/twilight.nvim/issues/20) ([04d0f27](https://github.com/folke/twilight.nvim/commit/04d0f2792f84a1135bb0db86a4e4942eb9f67aad))
* added config option to dim inactive windows. Fixes [#5](https://github.com/folke/twilight.nvim/issues/5) ([35b0c01](https://github.com/folke/twilight.nvim/commit/35b0c01fb8a978531fa648d156ce99fd1f97c4c9))
* allow disabling treesitter ([8ab43c0](https://github.com/folke/twilight.nvim/commit/8ab43c0fdc246fdd82382d7ef4ec383f8f755ef9))
* initial version ([16f6a9a](https://github.com/folke/twilight.nvim/commit/16f6a9ada7c26c46e1e3096204498a03642d5af0))


### Bug Fixes

* add setup to init ([f5718dd](https://github.com/folke/twilight.nvim/commit/f5718ddbd9bba3184493a4fc11cd38342966904f))
* calc hl group on enable and on ColorScheme. no need to do during setup ([9c9adb9](https://github.com/folke/twilight.nvim/commit/9c9adb93c6fcacb7f87fd14b5f3c87006f99b3c0))
* properly dim all windows when Twilight is enabled ([30ef601](https://github.com/folke/twilight.nvim/commit/30ef601f36db80dc5daa46c89a66fe4d84176c47))
* remove dependency on nvim-treesitter ([9531b8e](https://github.com/folke/twilight.nvim/commit/9531b8ea93869f5feac63cb154c325ea603c5392))
* set extmark prio to 10000 to prevent other plugins to bleed through the dimming. Fixes [#4](https://github.com/folke/twilight.nvim/issues/4) ([0ac66e8](https://github.com/folke/twilight.nvim/commit/0ac66e85f2d8d7c6eb7d263ffc39fa19b56d882f))
* use pcall for treesitter check. Fixes [#7](https://github.com/folke/twilight.nvim/issues/7) ([f722fa2](https://github.com/folke/twilight.nvim/commit/f722fa24411d0423a3d2daf6f1b5f8b11df9db2b))


lua/twilight/util.lua
local M = {}

function M.get_hl(name)
	local ok, hl = pcall(vim.api.nvim_get_hl_by_name, name, true)
	if not ok then
		return
	end
	for _, key in pairs({ "foreground", "background", "special" }) do
		if hl[key] then
			hl[key] = string.format("#%06x", hl[key])
		end
	end
	return hl
end

function M.hex2rgb(hex)
	hex = hex:gsub("#", "")
	return tonumber("0x" .. hex:sub(1, 2)), tonumber("0x" .. hex:sub(3, 4)), tonumber("0x" .. hex:sub(5, 6))
end

function M.rgb2hex(r, g, b)
	return string.format("#%02x%02x%02x", r, g, b)
end

function M.darken(hex, amount)
	local r, g, b = M.hex2rgb(hex)
	return M.rgb2hex(r * amount, g * amount, b * amount)
end

function M.blend(hex, bg_hex, alpha)
	local fg = { M.hex2rgb(hex) }
	local bg = { M.hex2rgb(bg_hex) }

	local blendChannel = function(i)
		local ret = (alpha * fg[i] + ((1 - alpha) * bg[i]))
		return math.floor(math.min(math.max(0, ret), 255) + 0.5)
	end

	return M.rgb2hex(blendChannel(1), blendChannel(2), blendChannel(3))
end

function M.is_dark(hex)
	local r, g, b = M.hex2rgb(hex)
	local lum = (0.299 * r + 0.587 * g + 0.114 * b) / 255
	return lum <= 0.5
end

function M.log(msg, hl)
	vim.api.nvim_echo({ { "Twilight: ", hl }, { msg } }, true, {})
end

function M.warn(msg)
	M.log(msg, "WarningMsg")
end

function M.error(msg)
	M.log(msg, "ErrorMsg")
end

return M


lua/twilight/view.lua
local config = require("twilight.config")

local M = {}

local ns = vim.api.nvim_create_namespace("twilight")

M.enabled = false

function M.enable()
  if not M.enabled then
    config.colors()
    M.enabled = true
    -- HACK: use defer_fn since TreeSitter is doing something that breaks the dimmers
    vim.cmd([[
        augroup Twilight
          autocmd!
          autocmd BufWritePost,CursorMoved,CursorMovedI,WinScrolled * lua require("twilight.view").update()
          autocmd WinEnter * lua require("twilight.view").on_win_enter()
          autocmd BufWritePost * lua vim.defer_fn(function()require("twilight.view").update()end, 0)
          autocmd ColorScheme * lua require("twilight.config").colors()
        augroup end]])
    M.started = true
    M.on_win_enter()
  end
end

function M.on_win_enter()
  local current = vim.api.nvim_get_current_win()
  for _, win in ipairs(vim.api.nvim_list_wins()) do
    if win ~= current then
      if config.options.dimming.inactive then
        local from, to = M.get_visible(win)
        for i = from, to do
          M.dim(vim.api.nvim_win_get_buf(win), i)
        end
      else
        M.update(win)
      end
    end
  end
  M.update()
end

function M.disable()
  if M.enabled then
    M.enabled = false
    pcall(vim.cmd, "autocmd! Twilight")
    pcall(vim.cmd, "augroup! Twilight")
    for _, buf in pairs(vim.api.nvim_list_bufs()) do
      if vim.api.nvim_buf_is_valid(buf) then
        M.clear(buf)
      end
    end
  end
end

function M.toggle()
  if M.enabled then
    M.disable()
  else
    M.enable()
  end
end

function M.clear(buf, from, to)
  from = from or 0
  to = to or -1
  if from < 0 then
    from = 0
  end
  vim.api.nvim_buf_clear_namespace(buf, ns, from, to)
end

-- TODO: fooo
function M.dim(buf, lnum)
  -- use extmarks directly so we can set the priority
  -- do a pcall instead to prevent spurious errors at the end of the doc
  pcall(vim.api.nvim_buf_set_extmark, buf, ns, lnum, 0, {
    end_line = lnum + 1,
    end_col = 0,
    hl_group = "Twilight",
    hl_eol = true,
    priority = 10000,
  })
  -- vim.api.nvim_buf_add_highlight(buf, ns, "Twilight", lnum, 0, -1)
end

function M.range(node)
  local from, _, to, _ = node:range()
  return from, to
end

function M.get_node(buf, line)
  local lines = vim.api.nvim_buf_get_lines(buf, line, line + 1, false)
  local col = lines[1] and (#lines[1] - 1) or 0
  if col < 0 then
    col = 0
  end

  local parser = vim.treesitter.get_parser(buf)
  local ret
  parser:for_each_tree(function(tree)
    if ret then
      return
    end
    local root = tree:root()
    if root then
      local node = root:descendant_for_range(line, col, line, col)
      local parent = node:parent()
      while parent and (parent:start() == line or parent:end_() == line) do
        node = parent
        parent = node:parent()
      end
      if node and node ~= root then
        ret = node
      end
    end
  end)
  return ret
end

function M.is_empty(buf, line)
  local lines = vim.api.nvim_buf_get_lines(buf, line, line + 1, false)
  if vim.fn.trim(lines[1]) == "" then
    return true
  end
  return false
end

function M.expand(buf, from, to, line)
  if line < 1 then
    return from, to
  end
  if M.is_empty(buf, line) then
    return math.min(from, line), math.max(to, line)
  end
  local node = M.get_node(buf, line)
  if not node then
    return from, to
  end
  local f, t = M.range(node)
  if f < from then
    from = f
  end
  if t > to then
    to = t
  end
  return from, to, node
end

function M.get_expand_root(node, opts)
  opts = opts or {}
  local root
  while node do
    if config.expand[node:type()] then
      if opts.minimal_scope then
        return node
      end
      root = node
    end
    node = node:parent()
  end
  return root
end

function M.get_context(buf, line)
  if config.options.treesitter and pcall(vim.treesitter.get_parser, buf) then
    local node = M.get_node(buf, line)
    local root = M.get_expand_root(node, config.options.expand_opts)
    if root then
      local from, to = M.range(root)
      return from + 1, to + 2
    end
    local from, to = M.expand(buf, line, line, line)

    while to - from < config.options.context do
      local pf, pt, pnode = M.expand(buf, from, to, from - 1)
      local nf, nt, nnode = M.expand(buf, from, to, to + 1)

      local mp = math.max(line - pf, pt - line)
      local mn = math.max(line - nf, nt - line)

      if (mp < mn) and not (from == pf and to == pt) then
        from = pf
        to = pt
        if pnode and M.get_expand_root(pnode) then
          break
        end
      elseif not (from == nf and to == nt) then
        from = nf
        to = nt
        if nnode and M.get_expand_root(nnode) then
          break
        end
      else
        break
      end
    end

    return from + 1, to + 2
  end

  local from = line - math.floor(config.options.context / 2)
  local to = line + math.floor(config.options.context / 2)

  local lcount = vim.api.nvim_buf_line_count(buf)

  if to > lcount then
    to = lcount
  end
  if from < 1 then
    from = 1
  end
  while from > 0 and not M.is_empty(buf, from) do
    from = from - 1
  end

  while to < lcount and not M.is_empty(buf, to) do
    to = to + 1
  end

  return from + 1, to + 2
end

function M.is_valid_buf(buf)
  -- Skip special buffers
  local buftype = vim.api.nvim_buf_get_option(buf, "buftype")
  if buftype ~= "" then
    return false
  end
  local filetype = vim.api.nvim_buf_get_option(buf, "filetype")
  if vim.tbl_contains(config.options.exclude, filetype) then
    return false
  end
  return true
end

function M.update(win)
  win = win or vim.api.nvim_get_current_win()

  if not M.enabled or not vim.api.nvim_win_is_valid(win) then
    return false
  end
  local buf = vim.api.nvim_win_get_buf(win)
  if not M.is_valid_buf(buf) then
    return
  end
  local cursor = vim.api.nvim_win_get_cursor(win)
  local from, to = M.get_context(buf, cursor[1] - 1)

  local dimmers = {}
  M.focus(win, from, to, dimmers)

  for _, other in ipairs(vim.api.nvim_list_wins()) do
    if other ~= win and vim.api.nvim_win_get_buf(other) == buf then
      M.focus(other, from, to, dimmers)
    end
  end

  M.clear(buf, from - 1, to - 1)

  for lnum, _ in pairs(dimmers) do
    M.dim(buf, lnum - 1)
  end
end

function M.get_visible(win)
  local info = vim.fn.getwininfo(win)
  return info[1].topline, info[1].botline + 1
end

function M.focus(win, from, to, dimmers)
  if not vim.api.nvim_win_is_valid(win) then
    return
  end

  local topline, botline = M.get_visible(win)

  for l = topline, botline do
    if l < from or l >= to then
      dimmers[l] = true
    end
  end
end

return M


lua/twilight/config.lua
local util = require("twilight.util")
local M = {}

--- @class TwilightOptions
local defaults = {
  dimming = {
    alpha = 0.25, -- amount of dimming
    -- we try to get the foreground from the highlight groups or fallback color
    color = { "Normal", "#ffffff" },
    term_bg = "#000000", -- if guibg=NONE, this will be used to calculate text color
    inactive = false, -- when true, other windows will be fully dimmed (unless they contain the same buffer)
  },
  context = 10, -- amount of lines we will try to show arounc the current line
  node_context = 0, -- amount of extra nodes to show. Only node types mention in expand are considered
  treesitter = true, -- use treesitter when available for the filetype
  -- treesitter is used to automatically expand the visible text,
  -- but you can further control the types of nodes that should always be fully expanded
  expand = { -- for treesitter, we we always try to expand to the top-most ancestor with these types
    "function",
    "method",
    "table",
    "if_statement",
  },
  exclude = {}, -- exclude these filetypes
}

--- @type TwilightOptions
M.options = {}
M.types = {}
M.expand = {}

function M.setup(options)
  M.options = vim.tbl_deep_extend("force", {}, defaults, options or {})
  for _, value in pairs(M.options.expand) do
    M.expand[value] = true
  end
end

function M.colors()
  local fg = "#ffffff"
  for _, color in pairs(M.options.dimming.color) do
    if color:sub(1, 1) == "#" then
      fg = color
      break
    end
    local hl = util.get_hl(color)
    if hl and hl.foreground then
      fg = hl.foreground
      break
    end
  end
  local normal = util.get_hl("Normal")
  local bg = M.options.dimming.term_bg
  if normal then
    bg = normal.background or "NONE"
  end
  -- use terminal background in blend function if guibg is NONE
  local bg_hex = (bg ~= "NONE") and bg or M.options.dimming.term_bg
  local dimmed = util.blend(fg, bg_hex, M.options.dimming.alpha)
  vim.cmd("highlight! def Twilight guifg=" .. dimmed .. " guibg=" .. bg)
end

M.setup()

return M
